### PR TITLE
ASoC: intel: sof_sdw: Use a fixed DAI link id for AMP

### PR DESCRIPTION
**Pull request by bardliao**
Currently, we assign SoundWire DAI link id according to the order in
snd_soc_acpi_link_adr. It will varies between devices and we have to
modify topologies for the different DAI link ids. Fix a fixed DAI link id
can reduce the possibility of editing topologies.
This patch has the assumption that the codec order in snd_soc_acpi_link_adr
is always headset -> amp -> mic. We have to follow the order when we
add new link adrs in the future.

### CODE REVIEW

**Comment by libinyang** [comment]
https://github.com/thesofproject/sof/pull/4824 is using the dynamic BE id now. Please see https://github.com/thesofproject/sof/pull/4824/commits/6329628426de759e0aba43bce3906ae49ed35944

So either this PR or https://github.com/thesofproject/sof/pull/4824/commits/6329628426de759e0aba43bce3906ae49ed35944 should be merged.

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -895,6 +912,9 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 	if (codec_info_list[codec_index].ignore_pch_dmic)
 		*ignore_pch_dmic = true;
 
+	if (is_amp(codec_info_list[codec_index].part_id) && link_id < SDW_AMP_DAI_ID)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: it would be simpler to add a field in codec_info_list, so that we have all the information in a single table.

we could even unify the rt715 handling with different flags for smart amp and smart mic.

**Review by plbossart — CHANGES_REQUESTED** [review]

**Comment by plbossart** [comment]
> [thesofproject/sof#4824](https://github.com/thesofproject/sof/pull/4824) is using the dynamic BE id now. Please see [thesofproject/sof@6329628](https://github.com/thesofproject/sof/commit/6329628426de759e0aba43bce3906ae49ed35944)
> 
> So either this PR or [thesofproject/sof@6329628](https://github.com/thesofproject/sof/commit/6329628426de759e0aba43bce3906ae49ed35944) should be merged.

I don't think it makes sense to change the ID in the topology.

One main reason: it would prevent us from testing the 'noajck' option on platforms that do have a jack. The rule is that the machine driver can expose as many links as possible, and the topology *selects* from the links. If we change the topology to match the machine driver we are adding an exception to the rule and creating more problems.

It's really simpler if we assume that the machine driver has a pre-defined link numbering.

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -1137,7 +1142,7 @@ static int sof_card_dai_links_create(struct device *dev,
 		    group_generated[endpoint->group_id])
 			continue;
 
-		ret = create_sdw_dailink(card, dev, &be_id, links, sdw_be_num,
+		ret = create_sdw_dailink(card, dev, &link_id, links, sdw_be_num,
```
Comment (libinyang) [inline sound/soc/intel/boards/sof_sdw.c]: The below ```dev_err(dev, "failed to create dai link %d", be_id);``` should also be changed to link_id

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: Done, thanks.

**Comment by bardliao** [comment]
> Can you add some logic to check the sequence of "headphone - AMP - MIC". A error message is helpful to users

Yes, added.

File: sound/soc/intel/boards/sof_sdw.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > it would be simpler to add a field in codec_info_list, so that we have all the information in a single table.
> 
> we could even unify the rt715 handling with different flags for smart amp and smart mic.

Done

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -845,6 +860,7 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 	struct snd_soc_dai_link_component *codecs;
 	int cpu_dai_id[SDW_MAX_CPU_DAIS];
 	int cpu_dai_num, cpu_dai_index;
+	static int link_id = 0;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: can we try to avoid using a static here?

File: sound/soc/intel/boards/sof_sdw.c [hunk 4]
```diff
@@ -915,6 +915,14 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 	    link_id < SDW_AMP_DAI_ID)
 		link_id = SDW_AMP_DAI_ID;
 
+	/*
+	 * DAI ID is fixed at SDW_DMIC_DAI_ID for MICs to
+	 * keep sdw DMIC and HDMI setting static in UCM
+	 */
+	if (codec_info_list[codec_index].codec_type == SOF_SDW_CODEC_TYPE_MIC &&
+	    link_id < SDW_DMIC_DAI_ID && (sof_sdw_quirk & SOF_RT715_DAI_ID_FIX))
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: should we remove this quirk now and use only the codec_type?

File: sound/soc/intel/boards/sof_sdw.c [hunk 5]
```diff
@@ -920,7 +907,7 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 	 * keep sdw DMIC and HDMI setting static in UCM
 	 */
 	if (codec_info_list[codec_index].codec_type == SOF_SDW_CODEC_TYPE_MIC &&
-	    link_id < SDW_DMIC_DAI_ID && (sof_sdw_quirk & SOF_RT715_DAI_ID_FIX))
+	    link_id < SDW_DMIC_DAI_ID)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: I would do this directly in the previous patch, and here only cleanup the quirks

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > can we try to avoid using a static here?

I used static is because create_sdw_dailink() will be called multiple time and I want to keep the value of link_id. But I don't like static inside a function either. Using link_id as an argument of create_sdw_dailink() now.

File: sound/soc/intel/boards/sof_sdw.c (hunk 5 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > I would do this directly in the previous patch, and here only cleanup the quirks

Done

File: sound/soc/intel/boards/sof_sdw.c [hunk 6]
```diff
@@ -559,6 +560,12 @@ static int get_sdw_dailink_info(const struct snd_soc_acpi_link_adr *links,
 		if (codec_index < 0)
 			return codec_index;
 
+		if (codec_info_list[codec_index].codec_type < _codec_type)
+			dev_warn(dev,
+				 "adr is out of order, please make sure the adr order is jack -> amp -> mic order\n");
+
+		_codec_type = codec_info_list[codec_index].codec_type;
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_sdw.c:577]: _codec_type  will be overridden by "int _codec_type = SOF_SDW_CODEC_TYPE_JACK;"

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:577]: > _codec_type will be overridden by "int _codec_type = SOF_SDW_CODEC_TYPE_JACK;"

Thanks, fix it now.

**Comment by bardliao** [comment]
> I am in bad status and can't figure out link_id, link_index now. 

`link_index` is the index of `dai_links`, `link_id` is the `be_id` argument of init_dai_link() and it will be `dai_links->id`. BTW, IMHO, it will be more intuitive if we rename `be_id` to `link_id` or just `id` in init_dai_link().

> Can you test on all sdw platforms to confirm the patch ? thanks.

I think CI will do it for me. I will check the result once it is finished.

**Comment by bardliao** [comment]
> > Can you test on all sdw platforms to confirm the patch ? thanks.
> 
> I think CI will do it for me. I will check the result once it is finished.

The only fail on sdw device is https://sof-ci.01.org/linuxpr/PR3203/build6565/devicetest/?model=ADLP_RVP_SDW&testcase=verify-kernel-boot-log. But it seems not be an audio issue, and I checked the dmesg
```
[    7.301601] kernel: sof_sdw sof_sdw: create dai link SDW0-Playback, id 0
[    7.301789] kernel: sof_sdw sof_sdw: create dai link SDW0-Capture, id 1
[    7.301801] kernel: sof_sdw sof_sdw: create dai link dmic01, id 2
[    7.301813] kernel: sof_sdw sof_sdw: create dai link dmic16k, id 3
[    7.301842] kernel: sof_sdw sof_sdw: create dai link iDisp1, id 4
[    7.301866] kernel: sof_sdw sof_sdw: create dai link iDisp2, id 5
[    7.301889] kernel: sof_sdw sof_sdw: create dai link iDisp3, id 6
[    7.301913] kernel: sof_sdw sof_sdw: create dai link iDisp4, id 7
[    7.301936] kernel: sof_sdw sof_sdw: create dai link SSP2-BT, id 8
```
It is exactly the same as the PASS case on another PR's pr-device-test result. Besides, all other test cases are PASS.
So I think my PR should be fine.

File: sound/soc/intel/boards/sof_sdw.c [hunk 7]
```diff
@@ -1137,23 +1142,20 @@ static int sof_card_dai_links_create(struct device *dev,
...
 			return ret;
 		}
 	}
 
-	/* non-sdw DAI follows sdw DAI */
-	link_id = be_id;
-
 	/* get BE ID for non-sdw DAI */
-	be_id = get_next_be_id(links, be_id);
+	be_id = get_next_be_id(links, link_id);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: can we rename this variable 'link_index' to be consistent with the rest of the uses where it's a pointer?
same comment for line 1152

File: sound/soc/intel/boards/sof_sdw.c [hunk 8]
```diff
@@ -409,6 +411,7 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 		.dai_name = "rt711-aif1",
 		.init = sof_sdw_rt711_init,
 		.exit = sof_sdw_rt711_exit,
+		.codec_type = SOF_SDW_CODEC_TYPE_JACK,
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:397]: in the case of RT711, we may want to be careful: it also has local microphone support, and more generally when we move to SDCA parts we can have devices with multiple Functions. So I wonder if we should treat the 'codec_type' as 'function_mask' to allow a device to have multiple functions.

We can also revisit this later since we would also need to have multiple dai names so there's still work to be done here.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:397]: > in the case of RT711, we may want to be careful: it also has local microphone support, and more generally when we move to SDCA parts we can have devices with multiple Functions. So I wonder if we should treat the 'codec_type' as 'function_mask' to allow a device to have multiple functions.
> 
> We can also revisit this later since we would also need to have multiple dai names so there's still work to be done here.

Supporting  multiple functions will be a problem indeed. I don't think current sof_sdw can support multiple functions. We specific a single `.dai_name` for each codec in codec_info_list[]. So, rt711 or rt711-sdca can only be a headset for now.

**Comment by bardliao** [comment]
> minor nit-pick and question below.
> 
> I think we should also allow the user to specify a quirk, as we did with bytcr_rt5640, that would allow us to solve user problems without waiting for a patched kernel to be available.

Sorry @plbossart what bytcr_rt5640 quirk you referring to? We do allow user to overwrite the quirk by module param. Or do you think we miss some quick? Allowing user to specific DAI link ids for example?

**Comment by plbossart** [comment]
> > minor nit-pick and question below.
> > I think we should also allow the user to specify a quirk, as we did with bytcr_rt5640, that would allow us to solve user problems without waiting for a patched kernel to be available.
> 
> Sorry @plbossart what bytcr_rt5640 quirk you referring to? We do allow user to overwrite the quirk by module param. Or do you think we miss some quick? Allowing user to specific DAI link ids for example?

Discard this comment, I didn't recall we already had a quirk module parameter :-)

**Review by plbossart — DISMISSED** [review]

File: sound/soc/intel/boards/sof_sdw.c [hunk 9]
```diff
@@ -1157,7 +1176,7 @@ static int sof_card_dai_links_create(struct device *dev,
 					 sdw_cpu_dai_num, cpus, adr_link,
 					 &cpu_id, group_generated,
 					 codec_conf, codec_conf_count,
-					 &codec_conf_index,
+					 &be_id, &codec_conf_index,
 					 &ignore_pch_dmic);
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw.c:1156]: 15 parameters... Can we use a structure? Unrelated, of course...

File: sound/soc/intel/boards/sof_sdw.c [hunk 10]
```diff
@@ -404,6 +404,7 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 		.direction = {true, true},
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_sdw.c:376]: Commit message needs a bit of work:
- "It will varies between" -> "It will vary between"
- "Fix a fixed DAI link can reduce .. possibility of editing topologies" -> ? not sure what this means

File: sound/soc/intel/boards/sof_sdw.c [hunk 11]
```diff
@@ -906,6 +921,10 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 	if (codec_info_list[codec_index].ignore_pch_dmic)
 		*ignore_pch_dmic = true;
 
+	if (codec_info_list[codec_index].codec_type == SOF_SDW_CODEC_TYPE_AMP &&
+	    *link_id < SDW_AMP_DAI_ID)
+		*link_id = SDW_AMP_DAI_ID;
+
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_sdw.c:906]: I guess this is ok, but  a little difficult to follow. So if link_id is larger than SDW_AMP_DAI_ID, then it's not fixed anymore. Should this be an error?

File: sound/soc/intel/boards/sof_sdw.c [hunk 12]
```diff
@@ -841,7 +841,7 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 }
 
 static int create_sdw_dailink(struct snd_soc_card *card,
-			      struct device *dev, int *be_index,
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c:844]: typo in the commit message " Rename them to link_index to make it less confusing" or "Rename them to link_index to be consistent"

File: sound/soc/intel/boards/sof_sdw.c (hunk 10 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c:376]: Also while you're at it, can you change the first line to something like "Currently, we assign SoundWire DAI link id according to the order in the link address table or array"

File: sound/soc/intel/boards/sof_sdw.c [hunk 13]
```diff
@@ -1048,7 +1053,7 @@ static int sof_card_dai_links_create(struct device *dev,
 	bool group_generated[SDW_MAX_GROUPS];
 	int ssp_codec_index, ssp_mask;
 	struct snd_soc_dai_link *links;
-	int num_links, link_id = 0;
+	int num_links, link_index = 0;
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c:1051]: @bardliao did you miss changing the link_id name on line 1169?

File: sound/soc/intel/boards/sof_sdw.c (hunk 11 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c:906]: what happens otherwise? is it legal? or are you leaving that at the mercy of topology changes?

File: sound/soc/intel/boards/sof_sdw.c [hunk 14]
```diff
@@ -984,17 +984,6 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 	return 0;
 }
 
-/*
- * DAI link ID of SSP & DMIC & HDMI are based on last
- * link ID used by sdw link. Since be_id may be changed
- * in init func of sdw codec, it is not equal to be_id
- */
-static inline int get_next_be_id(struct snd_soc_dai_link *links,
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c:989]: I'm not sure how this works @bardliao . This part of the code is already so confusing so I'm not saying its incorrect just a bit hard to follow. The DAI link ID is only set conditionally in create_sdw_link, right? WHat happens in case it isnt set ? Are you saying this function was unnecessary even before? In that case, this patch should precede the patch that sets a fixed DAI link ID right?

File: sound/soc/intel/boards/sof_sdw.c [hunk 15]
```diff
@@ -569,6 +570,12 @@ static int get_sdw_dailink_info(const struct snd_soc_acpi_link_adr *links,
 		if (codec_index < 0)
 			return codec_index;
 
+		if (codec_info_list[codec_index].codec_type < _codec_type)
+			dev_warn(dev,
+				 "adr is out of order, please make sure the adr order is jack -> amp -> mic order\n");
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c]: this comment has 3 "order";s. can you please change it to "Unexpected address table ordering. Expected order: jack -> amp(s) -> mic\n"

File: sound/soc/intel/boards/sof_sdw.c (hunk 11 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:906]: `*link_id >= SDW_AMP_DAI_ID` is still a legal case. The common case is that we have one soundwire headset and its link ID will be `0` for playback and `1` for capture. In this case, `*link_id` here is 2, and we don't need to change it. And we may have a second soundwire AMP which id will be `3` which is larger than `SDW_AMP_DAI_ID`.
The condition here is to only overwrite the first soundwire AMP's id if it is less than SDW_AMP_DAI_ID.
Also, I don't know will we have more than one soundwire headsets in a device in the future, but in such case, we don't want to change the soundwire AMP's id. 
In short, we can't change AMP's id if it is larger than `SDW_AMP_DAI_ID`, otherwise we will have duplicated ids.

File: sound/soc/intel/boards/sof_sdw.c (hunk 13 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1051]: > @bardliao did you miss changing the link_id name on line 1169?

Right, thanks for shooting this. You have a great eye, appreciated. `get_next_be_id()` will be removed by the ` ASoC: intel: sof_sdw: remove get_next_be_id` commit. That's why CI and I didn't notice that.

File: sound/soc/intel/boards/sof_sdw.c (hunk 14 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:989]: @ranj063 DAI link ID is **always** set by `init_dai_link()`. What we do in `create_sdw_dailink()` is to overwrite `*link_id` if needed. But we will always have a `*link_id` when calling `init_dai_link()` in `create_sdw_dailink()`.
We needed `get_next_be_id()` because we change link_id in codec's .init callback and we don't know what is the last link id. Therefore, we need `get_next_be_id()` to get the last DAI link ID.

File: sound/soc/intel/boards/sof_sdw.c (hunk 15 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: Done

File: sound/soc/intel/boards/sof_sdw.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:376]: > * "Fix a fixed DAI link can reduce .. possibility of editing topologies" -> ? not sure what this means

That was a typo of "Use a fixed DAI link..."

File: sound/soc/intel/boards/sof_sdw.c [hunk 16]
```diff
@@ -944,8 +944,12 @@ static int create_sdw_dailink(struct snd_soc_card *card,
...
 		}
 
-		if (*be_index >= sdw_be_num) {
-			dev_err(dev, " invalid be dai index %d", *be_index);
+		/*
+		 * We create sdw dai links at first stage, so link index should
+		 * not be larger than sdw_be_num
+		 */
+		if (*link_index >= sdw_be_num) {
+			dev_err(dev, " invalid dai link index %d", *link_index);
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_sdw.c]: remove leading white space  " invalid ...."

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: Done

File: sound/soc/intel/boards/sof_sdw.c (hunk 14 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c:989]: ahh I see. makes sense.

**Review by ranj063 — DISMISSED** [review]
looks good @bardliao

File: sound/soc/intel/boards/sof_sdw.c (hunk 11 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c:906]: Thanks @bardliao . I wonder should this be explained either in the commit message or a code comment. This is something anyone maintaining the code in the future has to understand.

File: sound/soc/intel/boards/sof_sdw.c (hunk 10 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c:376]: Hmm, this still seems wrong: "Use a fixed DAI link id can reduce the possibility of editing topologies."

This seems like a pretty important part of the patchset, why are we moving to a fixed DAI link id. I'm not 100% sure what the you mean here. Would this be correct:
"Use of a fixed DAI link id can reduce the need to edit topologies."

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:376]: suggested commit message edit:

Currently, we assign SoundWire DAI link id according to the order in
the link address table, with the assumption that the headset codec is listed first, then amplifiers and last capture devices. If the headset codec is not present in a platform, the dai link for amplifiers will be shifted, which can be handled in two ways
a) modify the topology to renumber the dailink changes
b) keep the dailink numbers constant in topology but also avoid the variations in the machine driver.

This patch adds support for option b), the dailink index for amplifiers and capture devices becomes fixed.

File: sound/soc/intel/boards/sof_sdw.c (hunk 11 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:906]: > Thanks @bardliao . I wonder should this be explained either in the commit message or a code comment. This is something anyone maintaining the code in the future has to understand.

Code comment is added.

File: sound/soc/intel/boards/sof_sdw.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:376]: @plbossart Commit message changed. Thanks for the suggestion.
